src/phpDocumentor/FilesystemFactory.php

```diff
@@ -0,0 +1,89 @@
+<?php
+/**
+ * This file is part of phpDocumentor.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @copyright 2010-2015 Mike van Riel<mike@phpdoc.org>
+ * @license   http://www.opensource.org/licenses/mit-license.php MIT
+ * @link      http://phpdoc.org
+ */
+
+namespace phpDocumentor;
+
+use \LogicException;
+use League\Flysystem\Adapter\Local;
+use League\Flysystem\Filesystem;
+use League\Flysystem\MountManager;
+use \DI\Container;
+
+/**
+ * Class FilesystemFactory
+ */
+final class FilesystemFactory
+{
+    /** @var MountManager */
+    private $mountManager;
+
+    /** @var Container */
+    private $container;
+
+    /** @var Filesystem */
+    private $filesystem;
+
+    /**
+     * @param MountManager $mountManager
+     * @param Container $container
+     */
+    public function __construct(MountManager $mountManager, Container $container)
+    {
+        $this->mountManager = $mountManager;
+        $this->container = $container;
+    }
+
+    /**
+     * Returns a Filesystem instance based on the scheme of the provided Dsn
+     *
+     * @param Dsn $dsn
+     * @return Filesystem
+     */
+    public function create(Dsn $dsn)
+    {
+        $dsnId = spl_object_hash($dsn);
+
+        if (! $this->setFilesystemWhenCached($dsnId)) {
+            if ($dsn->getScheme() === 'file') {
+
+                $path = $dsn->getPath();
```

Comment:
You can see an excellent example of why your implementation of the DSN shines here; you don't have to worry at the implementation side whether you would need to concatenate the host and path (as is the case with `file://); because the Value Object makes the translation it is only on one place instead of each caller